### DISCUSSION TITLE
Execution plane services - Stream CloudEvents via WebSocket or Polling via HTTP

### DISCUSSION THREAD

**Discussion by christian-kreuzberger-dtx** [keptn/keptn#2930] [General]
During discussion regarding execution plane services (= external Keptn-services  running outside of the Kubernetes Cluster where Keptn is installed) we discovered a problem:

> How can external Keptn-services subscribe to *all* CloudEvents?

# Status Quo

In general, by using the [keptn/distributor](https://github.com/keptn/keptn/tree/master/distributor) we abstract the connection to `nats-cluster`, and forward each CloudEvent coming from `nats-cluster` to the respective Keptn-service via a HTTP POST. Keptn-services do not need to know anything about NATS (just about CloudEvents).

This allows a Keptn-service to just listen on an HTTP port and process the CloudEvent is POSTed to this endpoint, e.g.:

https://github.com/keptn-sandbox/keptn-service-template-go/blob/634c4bc4bb9e45491d647d74514aceafe694cfd1/main.go#L183-L200

```
	// configure http server to receive cloudevents
	t, err := cloudeventshttp.New(
		cloudeventshttp.WithPort(env.Port),
		cloudeventshttp.WithPath(env.Path),
	)

	log.Fatalf("failed to start receiver: %s", c.StartReceiver(ctx, processKeptnCloudEvent))
```

## Handling external Keptn-services

By definition, an external Keptn-service does not have access to the "internal" Keptn-services, other than an exposed Keptn API. Therefore an external Keptn-service can not access `nats-cluster` at the moment.
Within the distributor we have implemented a polling mechanic that polls all *open* `*.triggered` Keptn CloudEvents from the Keptn API (e.g., `sh.keptn.event.deployment.triggered`), and forwards them to the Keptn-service.

However, this approach only works for `*.triggered` Keptn CloudEvents, but not for other Keptn CloudEvents like `sh.keptn.event.deployment.finished` or `sh.keptn.event.test.started`. 

## Do we actually need all events

Why would a microservice need `.started` and `.finished` events? Good question, here are some examples:
 
* Example 1: keptn-contrib/notification-service sends messages to chats (e.g., Slack) about those CloudEvents. For instance,  a `test.started` event would be nice for users to see as "Keptn: Tests for service ${SERVICE} in project ${PROJECT} have started at 12:34". In addition, the `test.finished` event would have a similar importance.
* Example 2: keptn-contrib/dynatrace-service forwards a `deployment.finished` event to the Dynatrace API. This aids visualization within Dynatrace UI, but also Dynatrace Root Cause Analysis can benefit from this information.
* Example 3: keptn-sandbox/statistics-service counts all events that are incoming. Here it also makes sense to see the number of `.triggered` vs. `.started` and `.finished` events.

# How to solve this problem

We initially discussed several possibilities, and want to present them here to trigger a discussion:

## Option 1 (by @christian-kreuzberger-dtx): Create an API Endpoint in keptn/api-service that streams all CloudEvents via a websocket

This option would allow the distributor to use the Keptn API to stream all CloudEvents via a websocket (= HTTP) connection,which is firewall friendly, and would require the distributor to only know the Keptn API token for connecting to the websocket.

The downside of this approach is scaling, as every external distributor would need to connect to the API which then opens a WebSocket connection.
However, right now the distributor polls events from the Keptn API via MongoDB datastore, and in my opinion this scales even less then the WebSocket approach.

## Option 2 (by @agrimmer): Use NATS WebSockets extension

In contrast to Option 1 where we self-implement the websocket, we could re-use an existing extension for NATS. The downside is that an external distributor would need to directly connect to the internal `nats-cluster`, which requires
* Hardening for our `nats-cluster` installation (so far we have never considered making `nats` publicly accessible)
* Credentials for accessing `nats-cluster` need to be given to the distributor

The huge benefit here would be that we rely on something that has already been prooven to work well.

**WARNING: This needs some research and a POC**

## Option 3 (no author): Poll all CloudEvents, not just `.triggered`

While right now we query all *open* `.triggered` event (there usually should not be too many *open*  `.triggered` events), we could implement a mechanism where we query all CloudEvents since a given timestamp.

This approach might already work via the existing Keptn API, but it might be quite troublesome from a performance perspective (imagine 10 external services querying all CloudEvents since time $x every 10 seconds).

# Discussion

None of the above approaches has been tried out with Keptn 0.8.x, so we definately need some POCs.
However, we also want opinions and other ideas, if possible.

Please comment, and raise your concerns or tell us which approach you like more.

**Comment by agrimmer** [comment]
As a mid-term goal, I also see the need that external services can subscribe to any event type.

However, for the short-term goal (i.e. Keptn 0.8.0), I would like to discuss what use cases do we want to support w.r.t. the split of the control- and execution-plane. 
More precisely, what services do we want to run in the execution-plane? I think that the `helm-service` and `jmeter-service` would be a good start. Or do we aim to also externally run e.g. the `dynatrace-service`?

This decision on the short-term goal decides the timeline for one of the proposed options.

**Reply by christian-kreuzberger-dtx** [reply]
I agree, `helm-service` and `jmeter-service` would be a perfect start.

Unfortunately, `helm-service` is one of the examples that triggered this discussion, as it needs to react on `.finished` events (at the moment), e.g.:
https://github.com/keptn/keptn/blob/ff5091b6204f1a3d953ee6fbfbcbc4e6d725190a/helm-service/main.go#L86-L88

https://github.com/keptn/keptn/blob/ff5091b6204f1a3d953ee6fbfbcbc4e6d725190a/helm-service/main.go#L92-L94

These two are actually also related to an issue that I'll create soon about reworking our CloudEvents for creating and deleting projects/services (e.g., post-service-create.triggered instead of service.create.finished)

**Reply by jetzlstorfer** [reply]
Another use-case we have currently in the [Litmus-service](https://github.com/keptn-sandbox/litmus-service) is that we are listening for `test-finished` events from the jmeter service and stop the Litmus tests. 
We have both services running in parallel as we need some load while we are executing the Litmus tests.

**Reply by johannes-b** [reply]
From a use-case perspective, let's concentrate on the `helm-service` and `jmeter-service` that run on the execution plane. From my current point of view, it is unlikely that "streaming" services like the `dynatrace-service` or `notification-service` would run on the execution plane. I see them closer at the control plane. 

Given that use-case perspective, we still cannot use the helm-service and litmus-service on the exec plane due to the mentioned limitations. 

I think that doing POCs for the above-mentioned options is time-consuming and will hinder us from getting the first version of the "control/execution plane split" to the Keptn users (and we won't get their feedback either). Based on their feedback we can then think of providing more advanced support, i.e., WebSocket, opening NATS. Nevertheless, we should not forget about the limitations:

**My proposal**

I propose (a kind of compromise) that we implement a `/event/finished/{eventType}` endpoint (*take a look at the /finished*) on the shipyard-controller that works like the endpoint for *triggered*  events, but just returns *finished* events. 
The implementation should be straightforward from my point of view.

 
As a result, we still can configure a distributor (on the exec plane) as follows: 

```
- name: PUBSUB_TOPIC
  value: 'sh.keptn.event.service.create.finished,sh.keptn.event.deployment.triggered,sh.keptn.event.release.triggered,sh.keptn.event.action.triggered,sh.keptn.event.service.delete.finished'
```

The distributor used the `/event/finished/{eventType}` endpoint for `.finished` events and the `/event/triggered/{eventType}` for .triggered events. Consequently, the `helm-service` can listen to `sh.keptn.event.service.create.finished`, or the litmus-service to `sh.keptn.event.test.finished` events.

**Comment by bacherfl** [comment]
The option I like best so far is to make use of the NATS WebSocket extension, since this option is likely to be the most battle-tested.

Another option would be to use a WebHook mechanism where URLs can be registered for certain events - this would increase interoperability since the subscribing services do not need to establish a websocket connection and rather provide an HTTP endpoint that should be called. This means however that we will also need to manage the credentials for the registered WebHook URLs within Keptn

**Reply by agrimmer** [reply]
The webhook mechanism is definitely a nice solution. However, the communication would then be from the control-plane to the execution-plane service, which we tried to avoid (e.g. because of firewall settings).

**Comment by johannes-b** [comment]
For Keptn 0.8.0, the helm-service will no longer take care of uploading the Helm chart of a service that is going to be onboarded. Hence, no need to listen to a `service.create.finished`. 
Additionally, we will have the limitation that a helm-service on the execution plane does not listen to a `delete.serviec.finished`. Consequently, deleting Helm releases on the execution plane will be a manual task. 

On the bottom line, 0.8.0 just allows listening to `triggered` events. 

Let`s re-open this discussion when we learned more about splitting the exec- from the control-plane.
